src/org.hdfgroup.object/hdf/object/CompoundDataFormat.java

```diff
@@ -32,7 +32,7 @@
      *
      * @return the number of members of the compound data object.
      */
-    public abstract int getMemberCount();
```

Comment:
If you're updating these interfaces, the public keyword is also redundant and unnecessary, though not really any need to waste time on it since it doesn't change anything.

Reply:
I wasn't sure about the public/private keyword, but I can update that as well. I decided to create an Attribute interface and am currently implementing that. Basically getting rid of AttributeDataset class and extending either ScalarDS or CompoundDS instead.

Reply:
I don't think an interface for Attributes is right. Just follow the Dataset pattern. Attribute should just be an abstract base class like it is now.

Reply:
Making it an interface allows me to follow the dataset pattern. An attribute is a dataset with properties. The properties come from the Attribute interface. Then I can still check if the object is an attribute, yet have the correct DS functions.

Reply:
A Dataset isn't an interface though and an Attribute class shouldn't be either. If there are some common methods, you could put them in an interface that the Attribute class implements (like MetaDataFormat, though this was sort of the point of MetaDataContainer), but an Attribute is an HObject and should act like one by either continuing to extend Dataset, or by separating itself from Dataset again and implementing a shared interface. The latter will just end up with a lot of repeated code though.

Reply:
No an implemented attribute is based on dataset - so H5ScalarAttr extends H5ScalarDS implements Attribute.

Reply:
Yes, we intentionally changed that so that Attributes would act like Datasets and share common code but the main common point between them is the DataFormat/CompoundDataFormat interfaces to make them work well with the TableViews. They're still conceptually different objects and Attributes don't need to extend Dataset, they just do it for code reuse convenience. What other non-dataset stuff do you need in an interface for Attributes?

Reply:
parent object and properties - five functions in total plus two variables. I like the way this is going, reduces the duplicate functions that are more complicated.

Reply:
Variables don't go inside interfaces anyway, so those won't be shared that way. We should leave the base Attribute class, otherwise it can start to get difficult to differentiate between Datasets and Attributes, which we'll need to do in some cases. If anything, it might be appropriate to have a MetaDataFormat interface that extends DataFormat, but otherwise turning Attributes into an interface is just going to end up messy.